libbpf-tools/sslsniff.bpf.c

```diff
@@ -0,0 +1,220 @@
+// SPDX-License-Identifier: (LGPL-2.1 OR BSD-2-Clause)
+// Copyright (c) 2023 Yusheng Zheng
+//
+// Based on sslsniff from BCC by Adrian Lopez & Mark Drayton.
+// 15-Aug-2023   Yusheng Zheng   Created this.
+#include <vmlinux.h>
+#include <bpf/bpf_core_read.h>
+#include <bpf/bpf_endian.h>
+#include <bpf/bpf_helpers.h>
+#include <bpf/bpf_tracing.h>
+#include "sslsniff.h"
+
+struct {
+    __uint(type, BPF_MAP_TYPE_PERF_EVENT_ARRAY);
+    __uint(key_size, sizeof(__u32));
+    __uint(value_size, sizeof(__u32));
+} perf_SSL_events SEC(".maps");
+
+#define BASE_EVENT_SIZE ((size_t)(&((struct probe_SSL_data_t *)0)->buf))
+#define EVENT_SIZE(X) (BASE_EVENT_SIZE + ((size_t)(X)))
+#define MAX_ENTRIES 10240
+
+#define min(x, y)                      \
+    ({                                 \
+        typeof(x) _min1 = (x);         \
+        typeof(y) _min2 = (y);         \
+        (void)(&_min1 == &_min2);      \
+        _min1 < _min2 ? _min1 : _min2; \
+    })
+
+struct {
+    __uint(type, BPF_MAP_TYPE_PERCPU_ARRAY);
```

Comment:
We have a compat layer in `compat.bpf.h` which abstracts this percpu array out, see `mountsnoop` for example.